Keep local copies of files in a separate mets:FLocat

kba opened the pull request:
One early decision that has haunted us for years now is that we have been using a single `mets:FLocat` for both the original URL of a `mets:file` and the local copy in the workspace we use for processing.

This PR tries to solve #323 by changing `OcrdFile` and the download logic in `Resolver` and `Workspace`:

- the original URL (`OcrdFile.url`) remains in `mets:FLocat[@LOCTYPE="URL"]/xlink:href`
- the local filename (`OcrdFile.local_filename`) will now be written to an additional `mets:Flocat[@LOCTYPE="OTHER"][@OTHERLOCTYPE="FILE"]/xlink:href`
- The contract of `Workspace.download_file` is that after calling it with `OcrdFile f`, `f` will have a `local_filename` attribute and that is what processors should use rather than the `url`.
- The logic of when to download and when to copy files in `Resolver.download_to_directory` and `Workspace.download_file` has been adapted accordingly.

The goal here is to make the OCR-D processing non-invasive. Currently, once you do `ocrd workspace find --download`, the original URL will be gone. With this PR, `ocrd workspace find --download` will add an additional `mets:Flocat` which can then be removed after processing is finished (to be compliant with the DFG Viewer METS profile) with `ocrd workspace find --undo-download`.

MehmedGIT commented:
When the files are downloaded there is an output with relative file path per line. When I did undo the downloading it returned just `None` on each line. Since the local path is removed, this makes sense, however, how should the general user interpret that output?

kba commented:
> Since the local path is removed, this makes sense, however, how should the general user interpret that output

Good point, I hadn't thought about that. Should be fixed in 0f26809e9

For the kant_aufklaerung_1784 test asset:

```
ocrd workspace find --download
OCR-D-IMG/INPUT_0017.tif
OCR-D-IMG/INPUT_0020.tif
OCR-D-GT-PAGE/PAGE_0017_PAGE.xml
OCR-D-GT-PAGE/PAGE_0020_PAGE.xml
OCR-D-GT-ALTO/PAGE_0017_ALTO.xml
OCR-D-GT-ALTO/PAGE_0020_ALTO.xml
```

and the reverse:

```
ocrd workspace find --undo-download
Removed local_filename OCR-D-IMG/INPUT_0017.tif
Removed local_filename OCR-D-IMG/INPUT_0020.tif
Removed local_filename OCR-D-GT-PAGE/PAGE_0017_PAGE.xml
Removed local_filename OCR-D-GT-PAGE/PAGE_0020_PAGE.xml
Removed local_filename OCR-D-GT-ALTO/PAGE_0017_ALTO.xml
Removed local_filename OCR-D-GT-ALTO/PAGE_0020_ALTO.xml
```

MehmedGIT commented:
Yes, that output is more convenient. Great!

bertsky commented on ocrd_models/ocrd_models/ocrd_file.py:
```diff
@@ -13,40 +14,35 @@ class OcrdFile():
         self.pageId = pageId
 
+        if local_filename:
+            self.local_filename = Path(local_filename)
```
Should have been `str(Path(local_filename))`, right? @kba

kba replied:
No, that was a conscious decision, though not optimally communicated. Changing the type helped me spot areas where `url` and `local_filename` were used interchangeably but I think it is a correct implementation that `local_filename` is a Path.

kba replied:
Reverted in #1182
